## Exercice 5


In [3]:
import numpy as np

#Load dataset
X_train = np.load('../data/classification/X_train.npy')
y_train = np.load('../data/classification/y_train.npy')
X_test = np.load('../data/classification/X_test.npy')
y_test = np.load('../data/classification/y_test.npy')

In [14]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVC

from sklearn.metrics import accuracy_score

svm_cls = SVC()
gradient_boosting_cls = GradientBoostingClassifier()

svm_cls.fit(X_train, y_train)
svm_cls_preds = svm_cls.predict(X_test)
svm_cls_accuracy = accuracy_score(y_test, svm_cls_preds)
print("SVM Classification Accuracy:", svm_cls_accuracy)

gradient_boosting_cls.fit(X_train, y_train)
gradient_boosting_cls_preds = gradient_boosting_cls.predict(X_test)
gradient_boosting_cls_accuracy = accuracy_score(y_test, gradient_boosting_cls_preds)
print("Gradient Boosting Classification Accuracy:", gradient_boosting_cls_accuracy)

SVM Classification Accuracy: 0.836
Gradient Boosting Classification Accuracy: 0.79


In [29]:
from sklearn.model_selection import GridSearchCV
svm_params = {
    'C': np.linspace(1, 2, 100),
    'degree' : [2,3,4,5]
}


# Perform grid search with cross-validation
svm_grid = GridSearchCV(svm_cls, svm_params)
svm_grid.fit(X_train, y_train)

# Evaluate the model on the test set
svm_cls_preds = svm_grid.predict(X_test)
svm_cls_accuracy = accuracy_score(y_test, svm_cls_preds)


print("SVM Classification Accuracy:", svm_cls_accuracy)
print("Best Parameters for SVM:", svm_grid.best_params_)



SVM Classification Accuracy: 0.78
Best Parameters for SVM: {'kernel': 'linear'}


After trying several parameter research on svm, we could not achieve better than 83.6, so we decided to try it on greadient boosting

In [32]:
gradient_boosting_params = {
    'n_estimators': [10, 50, 100, 150, 200, 175, 80],
    'learning_rate': np.logspace(-3, 0, 100)
}

# Perform grid search with cross-validation
gradient_boosting_grid = GridSearchCV(gradient_boosting_cls, gradient_boosting_params)
gradient_boosting_grid.fit(X_train, y_train)

# Evaluate the model on the test set
gradient_boosting_cls_preds = gradient_boosting_grid.predict(X_test)
gradient_boosting_cls_accuracy = accuracy_score(y_test, gradient_boosting_cls_preds)

print("Gradient Boosting Classification Accuracy:", gradient_boosting_cls_accuracy)
print("Best Parameters for Gradient Boosting:", gradient_boosting_grid.best_params_)

Gradient Boosting Classification Accuracy: 0.81
Best Parameters for Gradient Boosting: {'learning_rate': 0.8111308307896872, 'n_estimators': 175}


Gradient boosting did not achieve better and was really long to optimize

To achieve the accuracy asked, we will now train a Randomforest for the task

In [34]:
from sklearn.ensemble import RandomForestClassifier
rf_classifier = RandomForestClassifier()

param_grid = {
    'n_estimators': [50, 100, 150],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}

# Perform grid search with cross-validation
grid_search_rf = GridSearchCV(rf_classifier, param_grid)
grid_search_rf.fit(X_train, y_train)

# Print the best parameters and accuracy
print("Best Parameters:", grid_search_rf.best_params_)

# Evaluate the model on the test set
y_pred = grid_search_rf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Best Parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Accuracy: 0.854


Random tree forest seems better suited for this dataset as we achieve a 0.854 accuracy, with a low runtime